examples/sparse/gcnii.py

```diff
@@ -0,0 +1,166 @@
+import math
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity
+from torch.optim import Adam
+
+
+class GCNIIConvolution(nn.Module):
+    def __init__(self, in_size, out_size):
+        super().__init__()
+        self.out_size = out_size
+        self.weight = nn.Linear(in_size, out_size, bias=False)
+
+    ############################################################################
+    # (HIGHLIGHT) Take the advantage of DGL sparse APIs to implement the GCNII
+    # forward process.
+    ############################################################################
+    def forward(self, A_norm, H, H0, lamda, alpha, l):
+        beta = math.log(lamda / l + 1)
+
+        # Multiply a sparse matrix by a dense matrix.
+        H = A_norm @ H
+        H = (1 - alpha) * H + alpha * H0
+        H = (1 - beta) * H + beta * self.weight(H)
+        return H
+
+
+class GCNII(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size,
+        n_layers,
+        lamda,
+        alpha,
+        dropout=0.5,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.n_layers = n_layers
+        self.lamda = lamda
+        self.alpha = alpha
+
+        # The CGNII model.
+        self.layers = nn.ModuleList()
+        self.layers.append(nn.Linear(in_size, hidden_size))
+        for _ in range(n_layers):
+            self.layers.append(GCNIIConvolution(hidden_size, hidden_size))
+        self.layers.append(nn.Linear(hidden_size, out_size))
+
+        self.activation = nn.ReLU()
+        self.dropout = dropout
+
+    def forward(self, A_norm, feature):
+        H = feature
+        H = F.dropout(H, self.dropout, training=self.training)
+        H = self.layers[0](H)
+        H = self.activation(H)
+        H0 = H
+
+        # The CGNII convolution forward.
+        for i, conv in enumerate(self.layers[1:-1]):
+            H = F.dropout(H, self.dropout, training=self.training)
+            H = conv(A_norm, H, H0, self.lamda, self.alpha, i + 1)
+            H = self.activation(H)
+
+        H = F.dropout(H, self.dropout, training=self.training)
+        H = self.layers[-1](H)
+
+        return H
+
+
+def evaluate(model, A_norm, H, label, val_mask, test_mask):
+    model.eval()
+    logits = model(A_norm, H)
+    pred = logits.argmax(dim=1)
+
+    # Compute accuracy on validation/test set.
+    val_acc = (pred[val_mask] == label[val_mask]).float().mean()
+    test_acc = (pred[test_mask] == label[test_mask]).float().mean()
+    return val_acc, test_acc
+
+
+def train(model, g, A_norm, H):
+    label = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+    optimizer = Adam(model.parameters(), lr=0.01, weight_decay=5e-4)
+
+    loss_fcn = nn.CrossEntropyLoss()
+
+    for epoch in range(1500):
```

Comment:
50 epoch should be enough?
Same to GCN, 200 is unnecessary. https://github.com/dmlc/dgl/blob/master/examples/sparse/gcn.py